File: iosgl/iosgl/bt/material/MaterialBaseParam.m
//
//  MaterialBaseParam.m
//  iosgl
//
//  Created by zhao on 9/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "MaterialBaseParam.h"
#import "DynamicBaseConstItem.h"
#import "ConstItem.h"
#import "TexItem.h"
#import "Scene_data.h"
#import "TextureManager.h"

@implementation MaterialBaseParam
-(void)setData:(Material*)material ary:(NSArray<NSDictionary*>*)ary;
{
    MaterialBaseParam* this=self;
    this.material = material;
    this.dynamicConstList =  [NSMutableArray alloc];
    this.dynamicTexList = [NSMutableArray alloc];
    
    NSMutableArray<ConstItem*>* constList  = material.constList;
    NSMutableArray<TexItem*>* texList  = material.texList;
    
    for (int i = 0; i < ary.count; i++) {
        
        NSDictionary* obj = ary[i];
        int obj_type=[obj[@"type"]intValue];
        NSString* obj_name=(NSString*)obj[@"name"] ;
        if (obj_type == 0) {
            DynamicBaseTexItem* texItem   = [[DynamicBaseTexItem alloc]init];
            texItem.paramName = obj_name;
            for (int j = 0; j < texList.count; j++) {
                if (texItem.paramName == texList[j].paramName) {
                    texItem.target = texList[j];
                    break;
                }
            }
            int mipmap = 0;
            if (texItem.target) {
                mipmap = texItem.target.mipmap;
            }
            mipmap = 0;
            
            [[ TextureManager default]getTexture:[[Scene_data default]getWorkUrlByFilePath:obj[@"url"]] fun:^(NSObject * _Nonnull any) {
                texItem.textureRes=(TextureRes*)any;
            } wrapType:0 info:nil filteType:0 mipmapType:0];
            
            // [this.dynamicTexList addObject:texItem];
            
            
        } else {
            NSString* targetName = obj_name;
            ConstItem* target = nil;
            for (int j = 0; j < constList.count; j++) {
                if (targetName == constList[j].paramName0
                    || targetName == constList[j].paramName1
                    || targetName == constList[j].paramName2
                    || targetName == constList[j].paramName3) {
                    target = constList[j];
                    break;
                }
            }
            DynamicBaseConstItem* constItem = [[DynamicBaseConstItem alloc]init];
            [constItem setTargetInfo:target paramName:targetName type: obj_type ];
            if (obj_type== 1) {
                [constItem setCurrentVal:obj[@"x"]];
            } else if (obj_type== 2) {
                [constItem setCurrentVal:obj[@"x"] y:obj[@"y"]];
            } else {
                [constItem setCurrentVal:obj[@"x" ] y:obj[@"y"] z:obj[@"z"]];
            }
            [this.dynamicConstList addObject:constItem];
            
            
        }
    }
    
}
@end

File: android/SceneWorld001/app/src/main/java/z3d/material/MaterialBaseParam.java
package z3d.material;

import android.util.Log;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import z3d.base.MathCore;
import z3d.base.Scene_data;
import z3d.base.TexTuresBackFun;
import z3d.engine.GC;
import z3d.filemodel.TextureManager;

public class MaterialBaseParam extends GC {
    public Material material;
    public List<DynamicBaseTexItem> dynamicTexList ;
    public List dynamicConstList ;


    public void setData(   Material mater,  List<HashMap> ary )
    {
        this.material = mater;
        this.dynamicConstList = new ArrayList();
        this.dynamicTexList = new ArrayList<>();
        List<ConstItem> constList  = mater.constList;
        List<TexItem> texList  = mater.texList;
        for (int i = 0; i < ary.size(); i++) {
            try {

                HashMap obj = (HashMap) ary.get(i);

                int objType= (int) obj.get("type");
                if (objType== 0) {
                final    DynamicBaseTexItem texItem = new DynamicBaseTexItem();
                    texItem.paramName = (String) obj.get("name");
                    for (int j = 0; j < texList.size(); j++) {
                        if (texItem.paramName.equals(texList.get(j).paramName)) {
                            texItem.target = texList.get(j);
                            break;
                        }
                    }
                    int mipmap = 0;
                    if (texItem.target!=null) {
                        mipmap = texItem.target.mipmap;
                    }
                    mipmap = 0;
                    TextureManager.getInstance().getTexture(Scene_data.fileRoot+ obj.get("url"), new TexTuresBackFun() {
                        @Override
                        public void Bfun(TextureRes value) {
                            texItem.textureRes = value;
                        }
                    });
                    this.dynamicTexList.add(texItem);
                } else {
                    String targetName = (String)  obj.get("name");
                    ConstItem target=null ;
                    for (int j = 0; j < constList.size(); j++) {

                        if (targetName == constList.get(j).paramName0
                                || targetName == constList.get(j).paramName1
                                || targetName == constList.get(j).paramName2
                                || targetName == constList.get(j).paramName3) {

                            target = constList.get(j);
                            break;

                        }
                    }
                    DynamicBaseConstItem constItem = new DynamicBaseConstItem();
                    constItem.setTargetInfo(target, targetName, (int)objType);
                    List<Float> valArr=new ArrayList<>();
                    if (objType>= 1) {
                        valArr.add((float) obj.get("x"));
                    }
                    if (objType >= 2) {
                        valArr.add((float) obj.get("y"));
                    }
                    if (objType >= 3) {
                        valArr.add((float) obj.get("z"));
                    }
                    constItem.setCurrentVal(valArr);
                    this.dynamicConstList.add(constItem);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

}
